### PR TITLE
Fix memory overflow in CSR

### PR DESCRIPTION
**Pull request by Xiaohanjlll**
FIx: #1214 
Signed-off-by: Xiaohanjlll <hanx.xiao@intel.com>

### CODE REVIEW

File: os_stub/cryptlib_openssl/pk/x509.c [hunk 1]
```diff
@@ -266,8 +266,13 @@ bool libspdm_asn1_get_tag(uint8_t **ptr, const uint8_t *end, size_t *length,
         return false;
     }
 
-    ASN1_get_object((const uint8_t **)ptr, &obj_length, &obj_tag, &obj_class,
-                    (int32_t)(end - (*ptr)));
+    status = ASN1_get_object((const uint8_t **)ptr, &obj_length, &obj_tag, &obj_class,
+                    (int32_t)(end - (*ptr)));   
+    /* Either a primitive encoding with a valid tag and definite length, but the content octets won't fit into omax, or parsing failed. */
+    if (status == 0x80 || status == 0xa0){
```
Comment (jyao1) [inline os_stub/cryptlib_openssl/pk/x509.c:272]: please define macro for 0x80 and 0xa0.

Hardcode value is not allowed.

File: include/hal/library/cryptlib.h [hunk 2]
```diff
@@ -97,6 +97,8 @@
 #define LIBSPDM_CRYPTO_ASN1_TAG_CLASS_MASK 0xC0
 #define LIBSPDM_CRYPTO_ASN1_TAG_PC_MASK 0x20
 #define LIBSPDM_CRYPTO_ASN1_TAG_VALUE_MASK 0x1F
+#define LIBSPDM_CRYPTO_ASN1_PARSE_ERROR 0x80
+#define LIBSPDM_CRYPTO_ASN1_TOO_LONG 0xA0
```
Comment (jyao1) [inline include/hal/library/cryptlib.h:101]: No need to put to cryptlib.h

just put it to x509.c

File: os_stub/cryptlib_openssl/pk/x509.c [hunk 3]
```diff
@@ -21,6 +21,9 @@
 /*buffer size to store subject object*/
 #define MAX_SBUJECT_NAME_LEN 0x100
 
+#define LIBSPDM_CRYPTO_ASN1_PARSE_ERROR 0x80
```
Comment (steven-bellock) [inline os_stub/cryptlib_openssl/pk/x509.c:24]: How do you know this? Is this defined by ASN or OpenSSL? If it's defined by OpenSSL why aren't we using the OpenSSL #defines?

**Comment by jyao1** [comment]
NOTE: The case we are testing is to change  `LIBSPDM_DATA_TRANSFER_SIZE`  MACRO to `42` in spdm-emu.

Please ensure that configuration can work.

**Comment by Xiaohanjlll** [comment]
> NOTE: The case we are testing is to change `LIBSPDM_DATA_TRANSFER_SIZE` MACRO to `42` in spdm-emu.
> 
> Please ensure that configuration can work.

I check it now

**Comment by Xiaohanjlll** [comment]
> NOTE: The case we are testing is to change `LIBSPDM_DATA_TRANSFER_SIZE` MACRO to `42` in spdm-emu.
> 
> Please ensure that configuration can work.

LIBSPDM_DATA_TRANSFER_SIZE MACRO  changes to 42 and works normally.

**Comment by jyao1** [comment]
> LIBSPDM_DATA_TRANSFER_SIZE MACRO changes to 42 and works normally.

Sorry, please ignore that. Wrong input window.

File: os_stub/cryptlib_openssl/pk/x509.c (hunk 3 above)
Comment (Xiaohanjlll) [reply os_stub/cryptlib_openssl/pk/x509.c:24]: > How do you know this? Is this defined by ASN or OpenSSL? If it's defined by OpenSSL why aren't we using the OpenSSL #defines?

Sorry, I can't find a suitable macro in OpenSSL

Comment (jyao1) [reply os_stub/cryptlib_openssl/pk/x509.c:24]: 1) define a macro named "OPENSSL_ASN1_ERROR_MASK 0x80", and add link "https://man.openbsd.org/ASN1_get_object.3" to the comment for the MACRO.
2) use "ret & OPENSSL_ASN1_ERROR_MASK" in the code to judge if it is an error.

**Pull request merged**